### DISCUSSION TITLE
Type hinting **kwargs with Unpack in python 3.8 using typing_extensions.

### DISCUSSION THREAD

**Discussion by jgauth** [microsoft/pyright#7366] [Q&A] [answered]
I'd like to use a `TypedDict` and `Unpack` to type hint `**kwargs` like this;

```python
from typing import TypedDict
from typing_extensions import Unpack

class MyData(TypedDict):
    foo: int

def parse_kwargs(**kwargs: Unpack[MyData]):
    my_foo = kwargs["foo"]
```

However, I'm running python 3.8, which doesn't have `Unpack` in the `typing` module. Instead I'm trying to use `typing_extensions.Unpack`, but pyright is raising an error:

```python
error: Expected TypeVarTuple as type argument for Unpack
```
I'm using `typing_extensions==4.6.0` which should have the latest improvements to `Unpack` backported. https://typing-extensions.readthedocs.io/en/latest/#Unpack

Am I doing something wrong, or is this just not possible on python 3.8?

**Comment by erictraut** [comment, marked as answer]
That should work fine. I'm not able to repro the error that you're seeing.

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.8&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFCiSyKpoD6xAHjMSgM5Jg%2BmbLnwBVFAgCGVANa1aVADZTevKAFk45KTCkAKIgjKUaASgBctKDajAwYC5hR1apYsCjSQvYq1kA7lIgaLz6AFThgcGhThLScgDaWjp6ALqW1rYQcKz2YFAAvFDRIbyJAET5FWm0QA)

The only theory I have is that you're running a very old version of pyright that doesn't have support for [PEP 692](https://peps.python.org/pep-0692/). Which version of pyright are you running?

**Reply by jgauth** [reply]
Yes that was the issue. My CI was running `pyright 1.1.122`! Sorry for the bad question...

Thanks for the quick response, I really admire how you manage this project.

**Reply by jgauth** [reply]
Hi @erictraut , somewhat related followup question:

Subclassing and overriding a method that has `**kwargs` as an argument gives an error that I don't fully understand:
```python
from typing import TypedDict
from typing_extensions import Unpack

class Parent:
    def parse_kwargs(self, **kwargs) -> None:
        pass

class MyData(TypedDict):
    foo: int

class Child(Parent):
    def parse_kwargs(self, **kwargs: Unpack[MyData]) -> None: # <-- pyright error
        self.foo = kwargs["foo"]

```
Gives this error:
```
Method "parse_kwargs" overrides class "Parent" in an incompatible manner
  Parameter "**kwargs" has no corresponding parameter
```
Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.8&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFCiSyKpoD6xAHjMSgM5Jg%2BmbLnwBVFAgCGVANa1aVADZTevKAAUpIHjABctKEailiwKNJC9irWQHdtaXgAprS4ABooAKm-3HvACUUAC0AHxQAHKCxAbG8RaqvArKSVAAsnDkUjBSzkQIZJQ0gXHGwGBgepgodIoqalAAwgAWSEqkzlo6taWGxqbmlta2DiBOrsTuXr7%2B47zVEtJyANqZ2bkAusHhUTFlCVBuwAB0FWBQALxQc04rAETn95u0QA).

---
I was expecting this to work, based on how something similar _is_ allowed when we're not using `**kwargs` and `Unpack`:
```python
class Parent:
    def foo(self, x) -> None:
        pass

class Child(Parent):
    def foo(self, x: int) -> None: # <--- this is fine
        self.x = x
```
[playground link.](https://pyright-play.net/?pythonVersion=3.8&code=MYGwhgzhAEAKYCcCmA7ALgLgFDV9AJkgGbRED2ZAFBEiEQDTQAeAlNALQB80AcmSkmx5h0AA6QIWLKAnQAwgAsAliHyV4ydCyF5CJclRp1GTDNCVaO3PgJ0joRogDom0ALzMsQA)

Could you help me understand the issue here?

**Reply by erictraut** [reply]
The issue is that the method in the parent class accepts _any_ keyword arguments, but the overridden method accepts only a limited set of keyword arguments. It's OK for a child method to be less restrictive than the parent method, but the converse is not true. It's a violation of the Liskov Substitution Principle (LSP). 

I admit that the error message in this case is a bit misleading and could be improved.
